_posts/2025-05-05-date-histogram-optimizations.md

```diff
@@ -0,0 +1,216 @@
+As part of this blog, we will mainly focus on Date Histogram aggregations which is subtype of Bucket Aggregations
+
+## Why optimize Date Histograms: Use cases and benefits
+
+While performance optimizations are quite valuable in general, even more so for Date histograms which is widely used across various use cases. For example:
+
+- Analyzing web traffic patterns
+- Monitoring application metrics and logs
+- Visualizing sales trends
+- Tracking IoT sensor data
+
+Optimizing date histogram performance offers the following benefits:
+
+- **Faster analysis**: Reduced dashboard load times for time-series visualizations.
+- **Improved scalability**: Efficient handling of large data volumes without sacrificing performance.
+
+## How numeric data is stored in OpenSearch
+
+To address these inefficiency challenges, OpenSearch introduced optimizations that use the underlying index structure of date fields to significantly accelerate aggregation performance. To understand the optimizations better, let's review how OpenSearch stores numeric data like timestamps (using Lucene). Numeric data is stored in two main structures:
+
+1. **Document values (doc values)**: A columnar structure optimized for operations like sorting and aggregations. The traditional aggregation algorithm iterates over these values.
+2. **Index tree (BKD tree)**: A specialized index structure (a one-dimensional BKD tree for date fields) designed for fast range filtering. An index tree consists of inner nodes and leaf nodes. Values are stored only in leaf nodes, while inner nodes store the bounding ranges of their children. This structure allows efficient traversal in a sorted order to find documents within specific ranges.
+
+## How Aggregation works in OpenSearch
+
+![Aggregation execution flow](../assets/media/blog-images/2025-05-05-date-histogram-optimizations/Aggregation_Execution_Flow.png){:class="img-centered" width="650px"}
+
+By default, OpenSearch processes aggregations by first evaluating each query condition on every shard using Lucene, which builds iterators that identify matching document IDs. These iterators are then combined (e.g., using logical AND) to find documents that satisfy all query filters. The resulting set of matching document IDs is streamed to the aggregation framework, where each document flows through one or more aggregators. These aggregators use Lucene’s doc values (explained above) to efficiently retrieve the field values needed for computation (e.g., for calculating averages or counts). This streaming model is hierarchical—documents pass through a pipeline of aggregators, allowing them to be grouped into top-level and nested buckets simultaneously. For example, a document can first be bucketed by month, then further aggregated by HTTP status code within that month. This design enables OpenSearch to process complex, multi-level aggregations efficiently in a single pass over the matching documents.
+
+## Understanding the setup
```

Comment:
I feel it maybe better to organize 
`How numeric data is stored in OpenSearch`
`How Aggregation works in OpenSearch`
under this section